File: strings/05_bulls_n_cows/bulls_n_cows.py
from typing import Dict


class Solution:
    def get_hint(self, secret: str, guess: str) -> str:
        if not self._check_constraints(secret, guess):
            raise ValueError("Constraints violated")

        g = list(guess)
        bulls, cows = 0, 0
        digit_freq: Dict[str, int] = {}
        # First pass to find bulls
        for i in range(len(secret)):
            if secret[i] == g[i]:
                bulls += 1
                g[i] = '#'
            else:
                digit_freq[secret[i]] += 1
        # Second pass to find cows
        for i in range(len(g)):
            if (g[i] != '#') and (g[i] in digit_freq.keys()) and (digit_freq[g[i]] != 0):
                cows += 1
                digit_freq[g[i]] -= 1

        return f"{bulls}A{cows}B"

    def _check_constraints(self, s: str, g: str) -> bool:
        n = len(s)
        m = len(g)
        if n < 1 or n > 10**3:
            return False
        if m < 1 or m > 10**3:
            return False
        if n != m:
            return False

        for i in range(n):
            s_digit = int(s[i])
            g_digit = int(g[i])
            if not ((s_digit >= 0 and s_digit <= 9) or (g_digit >= 0 and g_digit <= 9)):
                return False

        return True

File: strings/05_bulls_n_cows/bulls_n_cows.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <cmath>
#include <stdexcept>

using namespace std;


class Solution {
public:
    string getHint(const string& secret, string guess) {
        if (!checkConstraints(secret, guess)) {
            throw std::runtime_error("Constraints violated");
        }
        int bulls = 0;
        int cows = 0;
        unordered_map<char, int> digit_freq;
        // First pass to find bulls
        for (int i = 0; i < secret.size(); i++) {
            if (secret[i] == guess[i]) {
                bulls++;
                guess[i] = '#';  // Mark this position as visited
            } else {
                digit_freq[secret[i]]++;
            }
        }
        // Second pass to find cows
        for (int i = 0; i < guess.size(); i++) {
            if (guess[i] != '#' && digit_freq.find(guess[i]) != digit_freq.end() && digit_freq[guess[i]] != 0) {
                cows++;
                digit_freq[guess[i]]--;  // Decrease the count of matched digits
            }
        }
        return to_string(bulls) + 'A' + to_string(cows) + 'B';
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const string& s, const string& g) {
        if (s.size() < 1 || s.size() > pow(10, 3)) return false;
        if (g.size() < 1 || g.size() > pow(10, 3)) return false;
        if (g.size() != s.size()) return false;

        for (int i = 0; i < s.size(); i++) {
            int s_digit = s[i] - '0';
            int g_digit = g[i] - '0';
            if (!((s_digit >= 0 && s_digit <= 9) && (g_digit >= 0 && g_digit <= 9))) {
                return false;
            }
        }
        return true;
    }
};
